nomt/src/beatree/ops/update/leaf_stage.rs

```diff
@@ -35,50 +37,95 @@ pub struct ChangedLeafEntry {
     pub inserted: Option<LeafNode>,
 
     // the separator of the next node.
-    pub next_separator: Option<Key>,
+    next_separator: Option<Key>,
+}
+
+pub enum LeavesTrackerEntry {
+    ChangedLeaf(ChangedLeafEntry),
+    UnchangedRange(SeparatorRange),
 }
 
 #[derive(Default)]
-pub struct LeafChanges {
-    // TODO: there's no real reason for this to be a BTreemap rather than a Vec. We push to the
-    // end always and drain from the left.
-    inner: BTreeMap<Key, ChangedLeafEntry>,
+pub struct LeavesTracker {
+    inner: BTreeMap<Key, LeavesTrackerEntry>,
+    // pending base received from the right worker which will be used as new base
+    pending_base: Option<(Key, LeafNode, Option<Key>)>,
     overflow_deleted: Vec<Vec<u8>>,
 }
 
-impl LeafChanges {
-    pub fn delete(&mut self, key: Key, pn: PageNumber, next_separator: Option<Key>) {
-        let entry = self.inner.entry(key).or_insert_with(|| ChangedLeafEntry {
-            deleted: None,
-            inserted: None,
-            next_separator: None,
-        });
-
-        entry.next_separator = next_separator;
+impl LeavesTracker {
+    pub fn delete_leaf(&mut self, key: Key, pn: PageNumber, next_separator: Option<Key>) {
+        let entry = self
+            .inner
+            .entry(key)
+            .or_insert(LeavesTrackerEntry::ChangedLeaf(ChangedLeafEntry {
+                deleted: None,
+                inserted: None,
+                next_separator,
+            }));
+
+        let changed_entry = match entry {
+            LeavesTrackerEntry::ChangedLeaf(changed_entry) => changed_entry,
+            LeavesTrackerEntry::UnchangedRange(_) => panic!("TODO"),
+        };
 
         // we can only delete a leaf once.
-        assert!(entry.deleted.is_none());
+        assert!(changed_entry.deleted.is_none());
 
-        entry.deleted = Some(pn);
+        changed_entry.deleted.replace(pn);
+        changed_entry.next_separator = next_separator;
     }
 
-    pub fn insert(&mut self, key: Key, node: LeafNode, next_separator: Option<Key>) {
-        let entry = self.inner.entry(key).or_insert_with(|| ChangedLeafEntry {
-            deleted: None,
-            inserted: None,
-            next_separator: None,
-        });
+    pub fn insert_leaf(&mut self, key: Key, node: LeafNode, next_separator: Option<Key>) {
+        if let Some(last_entry_next_separator) = self.get_last_next_separator() {
```

Comment:
naming: `last_next_separator` sounds like an oxymoron. any ideas for a clearer term?

Reply:
yes, it could for sure be better named. I could go for `higher_next_separator`